### PR TITLE
Feat/custom pk model identifier

### PR DESCRIPTION
**Pull request by HuiSF**
*Issue #, if available:*

#1426 

*Description of changes:*

- Deprecate `Model.getId`
  - Will make changes to `observeQuery` in DataStore which uses `getId()` in a separate PR
  - Required change in the API category will be followed up seperately
- Add `modelIdentifier` getter to `Model`
- Update unit test models
- Reconcile DataStore test models to `amplify_test` package
- Regenerate test models with codegen on this feature branch https://github.com/aws-amplify/amplify-codegen/tree/custom-pk-flutter (pending PR https://github.com/aws-amplify/amplify-codegen/pull/394)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_api/lib/src/graphql/paginated_result_impl.dart
Comment (dnys1) [inline packages/amplify_api/lib/src/graphql/paginated_result_impl.dart:49]: Why would this not be a list of model identifiers?

File: packages/amplify_datastore/example/lib/models/BelongsToChildExplicit.dart
Comment (dnys1) [inline packages/amplify_datastore/example/lib/models/BelongsToChildExplicit.dart:211]: These need `@override` annotations.

File: packages/amplify_datastore/example/lib/models/SimpleCustomType.dart
Comment (dnys1) [inline packages/amplify_datastore/example/lib/models/SimpleCustomType.dart:34]: This is a breaking change we should note

**Review by dnys1 — COMMENTED** [review]
I could be missing it - I don't see any models using a custom primary key

File: packages/amplify_api/lib/src/graphql/paginated_result_impl.dart
Comment (dnys1) [reply packages/amplify_api/lib/src/graphql/paginated_result_impl.dart:49]: Or why does it need it at all?

File: packages/amplify_datastore/example/lib/models/SimpleCustomType.dart
Comment (HuiSF) [reply packages/amplify_datastore/example/lib/models/SimpleCustomType.dart:34]: Hum I missed this part somehow, I don't this this is an expected change, let me double check, could be just the test model is outdated (generated with WIP codegen of other features)

Comment (ragingsquirrel3) [reply packages/amplify_datastore/example/lib/models/SimpleCustomType.dart:34]: I believe this is expected change in codegen that is specific to > 0.4.0, though possible we did not note this change properly. However, I think any breakage was already with 0.3.x - 0.4.x

**Review by ragingsquirrel3 — COMMENTED** [review]
LGTM, spoke offline about an integ test failure and want to resolve that before approving (or if non-issue with implementation and just tests, can be followup PR IMO).

File: packages/amplify_test/lib/test_models/Blog.dart
Comment (HuiSF) [reply packages/amplify_test/lib/test_models/Blog.dart:151]: By requirement, PK fields should be immutable within a "copy" factory/custructor/builder.

**Comment by HuiSF** [comment]
> LGTM, spoke offline about an integ test failure and want to resolve that before approving (or if non-issue with implementation and just tests, can be followup PR IMO).

The integ test failure should be caused by the issue that we patched in 0.4.1. I will update the PR branch.

**Comment by HuiSF** [comment]
> approved assuming integ test issue resolved w rebase

Thanks for the follow up @ragingsquirrel3 ! yes the integration tests all passed after merging the main branch.

File: packages/amplify_api/lib/src/graphql/paginated_result_impl.dart
Comment (HuiSF) [reply packages/amplify_api/lib/src/graphql/paginated_result_impl.dart]: Forgot to respond to this question. This dummy getter is added because `PaginatedResultImpl` extends from `Model`, added this following the original impl of `getId` in this class. Been syncing up with @ragingsquirrel3 regarding this part, this may be refactored when finalize custom PK support in API category.

**Review by dnys1 — APPROVED** [review]

**Pull request merged**
